### PR TITLE
feat: create config package for config validation

### LINKED ISSUE #2918: Change baseUrl during test (set to null) in order to visit local file instead of URL
### Current behavior:
We are currently implementing email confirmation test. 
We currently download the email file -> parse out the html content of it -> save it locally in fixture/. 

In order for `cypress.visit('path-to-local-file.html')` to load up the downloaded content, we need to change the `baseUrl` that was set in `cypress.json` to be empty. 

`Cypress.config('baseUrl', ' `');` doesn't switch `baseUrl` and keeps appending the path to local file to the original `baseUrl`, eg:  https://demo.test.com/fixtures/email-content.html

Also tried to follow https://docs.cypress.io/api/plugins/configuration-api.html#Usage but it also fails to switch out `baseUrl`. 

I couldn't find any workaround. 

One method I tried was NOT setting `baseUrl` in `cypress.json` and specify the main domain everywhere except for visiting local html file. But this also fails because we have a `before` hook that generates unique string for email and if you change the domain on `cy.visit()`, the `before` hook runs and the email file name won't match.

### Desired behavior:

`Cypress.config('baseUrl', ' ');` should switch baseUrl to none. 

### Steps to reproduce: (app code and test code)

Posting it as pseudo-code. Will fill in more details if needed.

**cypress.json**
```json
{
    "baseUrl": "https://demo.testing.com",
    "userAgent": "testing"
}
```

**test spec file**
```js
 var emailFileName;

    before(function () {
        cy.log('ran');
        cy.task('generateGuid').then(function ($guid) {
            guid = $guid;
            emailFileName = 'tester' + guid + '@test.com';
            cy.log(emailText);
        });
    });

it('Register user', function () {

        cy.visit('/login'); //-> yields https://demo.testing.com/login
        // Fill up registration info then click 'Register' button

        cy.wait(5000);
        cy.get(sendConfirmEmailButton).click({ force: true });

        // wait for 30s to verify email body
        cy.wait(20000);
        cy.downloadEmailHtml(emailFileName, 'confirmAccount');
    });

it('Visit email user', function () {
        // Change baseurl
        Cypress.config('baseUrl', ' ');
        // TODO visit/load the confirm account html file
        cy.visit(`cypress/fixtures/confirmAccount_${emailFileName}@test.com.html`);
        cy.get('#templateBody')
            .contains('Click the button below to confirm your account');
    });
```

### Versions

Cypress: 3.1.3
OS: Mac- mojave

### PR DESCRIPTION
**Pull request by emilyrohrbough**
- Closes #2918 

This change creates a `packages/config` package which provides the list of configuration options and their default values and validation functions. The majority of this work has been extracting the existing configuration logic out of the `packages/server` module so it could also be used by the `packages/driver` module during in-test configuration updates made via `Cypress.config`.

### User facing changelog

Add validation to in-test configuration updates made with `Cypress.config`.

### How has the user experience changed?

User will be shown an error message if their tests attempts to update the cypress configuration with an invalid value. Previously, the configuration would silently be updated and could cause un-expected behaviors or appear to be broken.

Example as seen in #2918, setting the baseUrl to an `' '` is not a valid baseUrl value. After this change the following would be displayed:

Screen Shot 2021-10-21 at 10 12 18 AM

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue or this PR been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?

### CODE REVIEW

**Comment by emilyrohrbough** [comment]
@chrisbreiding With the latest changes with better error messaging & code blocks for users:
```js
it('inline test config override throws error', () => {
  Cypress.config('baseUrl', 'null')
})
```
Screen Shot 2021-11-02 at 10 38 18 AM
```js
it('inline test config override throws error when executed from cy cmd', () => {
  cy.then(() => {
    Cypress.config('baseUrl', 'null')
  })
})
```
Screen Shot 2021-11-02 at 10 38 24 AM
```js
it('test config override throws error', { retries: '1' }, () => {})
```
Screen Shot 2021-11-02 at 10 38 36 AM
```js
describe('context config overrides throws error', { retries: '1' }, () => {
  it('runs', () => {})
})
```
Screen Shot 2021-11-02 at 10 38 46 AM
```js
describe('nested contexts overrides throws error at the correct line number', { retries: '1' }, () => {
  it('test override not reached', { baseUrl: 'not_an_http_url' }, () => {})
  it('does execute for each test', { baseUrl: 'not_an_http_url' }, () => {})
})
```
Screen Shot 2021-11-02 at 10 39 24 AM
```js
describe('nested contexts ', () => {
  describe('test override', () => {
    it('throws error at the correct line number', { baseUrl: 'not_an_http_url' }, () => {})
  })
})
```
Screen Shot 2021-11-02 at 10 39 32 AM

File: .gitignore
Comment (mjhenkes) [inline .gitignore:17]: I might be confused, but don't you want to ignore `lib`, not `src`?

Comment (emilyrohrbough) [reply .gitignore]: I think I fixed it this round. Theres are some expected scripts I missed the first round for local development: [`953dacd` (#18589)](https://github.com/cypress-io/cypress/pull/18589/commits/953dacd0e216d05082fa629e6da4fd2f520c6b5e)

File: system-tests/test/reporters_spec.js
Comment (jennifer-shehane) [inline system-tests/test/reporters_spec.js:164]: Why was this test expectation changed? Why are these tests touched in this PR?

Comment (emilyrohrbough) [reply system-tests/test/reporters_spec.js:164]: This test was updated to run locally with the same CLI flags Circle passes to produce the same colorized logs. I added the additional messages to indicate where the error occurred. I had updated logic for gathering the test override configuration so we could trace the line number of any configuration that fails validation. With this change, this test caught that the slowTestThreshold configuration was't being set. 

This tests should not have been changed:
```
expect(result.stdout.match(/passes quickly(.*)/)[1], 'when verifying "passes quickly" test time colors').contain('\u001b[33m')
```

Cleaned this change up here [`f0676e2` (#18589)](https://github.com/cypress-io/cypress/pull/18589/commits/f0676e2e7acdb9c9e6cfa9420987c8be374d6e98).

File: package.json
Comment (flotwig) [inline package.json:20]: (can't comment on the right line, but:) I think the "test" command also needs to be updated so the `config` package tests run in CI:

```
"test": "yarn lerna exec yarn test --scope cypress --scope \"'@packages/{config,electron,extension,https-proxy,launcher,net-stubbing,network,proxy,rewriter,runner,runner-shared,socket}'\"",
```

Right now I'm not sure they run in CI. I don't see them in `unit-tests`: https://app.circleci.com/pipelines/github/cypress-io/cypress/26077/workflows/af61706e-6d22-4391-bcf9-4c07335b7ac6/jobs/981791

(why do we have a `test-unit` and a `test` command? no good reason)

**Review by flotwig — CHANGES_REQUESTED** [review]
New errors look great, really happy this is getting close.

I think that the driver failures *are* actually related to the changes in this branch, though:

Compare the last failures in:

* develop: https://app.circleci.com/pipelines/github/cypress-io/cypress?branch=develop
* this branch: https://app.circleci.com/pipelines/github/cypress-io/cypress?branch=add-config-validation

In this branch, "src/cy/commands/navigation #page load errors can time out" failed in the last 5 reruns of the driver tests for `firefox` and `electron`, and also seems to have been failing for several commits before that (in those 2 browsers). It also appears to flakily fail in Chrome, which is strange.

But if you compare to `develop`, I don't see this test failing in there at all. [Dashboard also doesn't show it in `develop`](https://dashboard.cypress.io/projects/ypt4pf/analytics/flaky-tests?branches=%5B%7B%22value%22%3A%22develop%22%2C%22label%22%3A%22develop%22%7D%5D&browsers=%5B%5D&chartRangeMostCommonErrors=%5B%5D&chartRangeSlowestTests=%5B%5D&chartRangeTopFailures=%5B%5D&committers=%5B%5D&cypressVersions=%5B%5D&flaky=%5B%5D&operatingSystems=%5B%5D&runGroups=%5B%5D&specFiles=%5B%5D&status=%5B%7B%22label%22%3A%22Passed%22%2C%22value%22%3A%22PASSED%22%7D%2C%7B%22label%22%3A%22Failed%22%2C%22value%22%3A%22FAILED%22%7D%5D&tags=%5B%5D&timeInterval=WEEK&timeRange=%7B%22startDate%22%3A%222021-10-05%22%2C%22endDate%22%3A%222021-11-04%22%7D&viewBy=TEST_CASE), but does show it [flake in this branch](https://dashboard.cypress.io/projects/ypt4pf/analytics/flaky-tests?branches=%5B%7B%22label%22%3A%22add-config-validation%22%2C%22suggested%22%3Afalse%2C%22value%22%3A%22add-config-validation%22%7D%5D&browsers=%5B%5D&chartRangeMostCommonErrors=%5B%5D&chartRangeSlowestTests=%5B%5D&chartRangeTopFailures=%5B%5D&committers=%5B%5D&cypressVersions=%5B%5D&flaky=%5B%5D&operatingSystems=%5B%5D&runGroups=%5B%5D&specFiles=%5B%5D&status=%5B%7B%22label%22%3A%22Passed%22%2C%22value%22%3A%22PASSED%22%7D%2C%7B%22label%22%3A%22Failed%22%2C%22value%22%3A%22FAILED%22%7D%5D&tags=%5B%5D&timeInterval=WEEK&timeRange=%7B%22startDate%22%3A%222021-10-05%22%2C%22endDate%22%3A%222021-11-04%22%7D&viewBy=TEST_CASE)

**Comment by emilyrohrbough** [comment]
> I think that the driver failures are actually related to the changes in this branch

hmmm... Any insights on what might have impacted this?

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

File: circle.yml
Comment (emilyrohrbough) [reply circle.yml:967]: Assumed it was a typo given the `wait-for-circle-jobs` script parses on the commas: https://github.com/cypress-io/cypress/blob/develop/scripts/wait-on-circle-jobs.js#L170

File: packages/config/package.json
Comment (emilyrohrbough) [reply packages/config/package.json:8]: I messed with this a bit and this was worked for both local and prod tests/builds since typescript is compile `src` to `lib` when build is ran. Can you explain the piggy-backing concept a bit more? Does this work because the typescript files in `packages/launcher` & `packages/proxy` are being compiled to the same directory as the `.ts` files are located?

File: .vscode/extensions.json
Comment (emilyrohrbough) [reply .vscode/extensions.json]: How about this? [`b351eda` (#18589)](https://github.com/cypress-io/cypress/pull/18589/commits/b351edaf33a80d65dcf08ca893e250193179cc4a#diff-c16655a98a3ee89a7636a59c59a72b0e93649e3a1e947327cfc43a1336b4e912R3)

File: package.json
Comment (emilyrohrbough) [reply package.json:20]: good call: [`b351eda` (#18589)](https://github.com/cypress-io/cypress/pull/18589/commits/b351edaf33a80d65dcf08ca893e250193179cc4a#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R49)

File: .eslintignore
Comment (emilyrohrbough) [reply .eslintignore]: fixed; [`b351eda` (#18589)](https://github.com/cypress-io/cypress/pull/18589/commits/b351edaf33a80d65dcf08ca893e250193179cc4a#diff-a0fdacd2ade87c5238dde44378f574f7e123e669acdf8b740878b14b33b20275R37)

File: packages/driver/src/cypress/error_messages.ts
Comment (emilyrohrbough) [reply packages/driver/src/cypress/error_messages.ts]: fixed: [`b351eda` (#18589)](https://github.com/cypress-io/cypress/pull/18589/commits/b351edaf33a80d65dcf08ca893e250193179cc4a#diff-6c0564db103220b716eb75e4a31c25a8192ac068e816710298d2e14d80cf7499R254)

File: packages/config/package.json
Comment (chrisbreiding) [reply packages/config/package.json:8]: > Does this work because the typescript files in packages/launcher & packages/proxy are being compiled to the same directory as the .ts files are located?

Yeah, that's right. Is there any reason we can't do that here? I think we should because otherwise if you change a src file in `packages/config` in the course of development, it will be out-of-date unless you re-run the build script (or we'd need to set up a watch to keep the build up-to-date).

Comment (emilyrohrbough) [reply packages/config/package.json:8]: We could, but it was kind of confusing to have compiled & non-compiled code within the same directory. Several of the packages weren't cleaning their compiled files up correctly and I had to go through & delete them because some have .ts and .js files intermixed before the compiler is ran.

Comment (chrisbreiding) [reply packages/config/package.json:8]: Since it will only build for production, which only happens in CI, there wouldn't be compiled files alongside non-compiled files locally while developing. In which packages did you see compiled files? That should only happen if `build-prod` is run.

File: packages/config/src/index.js
Comment (mjhenkes) [reply packages/config/src/index.js:1]: @flotwig I agree, some of lodash is useful, but it would be nice to target just those functions over importing the entirety of lodash. I'm very much in favor of removing dependencies whenever we can, from both a bundle size and risk mitigation (either goofy changes, or security issues) 

i.e
```diff
-"lodash": "4.17.21"
+"lodash. camelCase": "^4.3.0",
```

**Review by mjhenkes — DISMISSED** [review]

File: packages/config/package.json
Comment (emilyrohrbough) [reply packages/config/package.json]: I had errors in CI with build-prod so I was running it locally to debug and noticed the behavior. I'm not totally stuck on this change, just felt it was clearer IMO. I removed it here: [`8bdfd57` (#18589)](https://github.com/cypress-io/cypress/pull/18589/commits/8bdfd57be4cc3115b94c4013a81d8b629f81d318)

**Review by mjhenkes — APPROVED** [review]

File: packages/driver/src/cy/testConfigOverrides.ts
Comment (chrisbreiding) [inline packages/driver/src/cy/testConfigOverrides.ts:46]: Just wrote a [better abstraction for this](https://github.com/cypress-io/cypress/pull/18587/files#diff-8ca2c866dca492abea845c9ef35963f7a539e0ad761cdca8b0ac0a1e408e27b0R135) in David's PR. Don't need to change it in this PR, since that would mean waiting for his to get merged first, but I wanted to note that this could be cleaned up in the future.

Comment (emilyrohrbough) [reply packages/driver/src/cy/testConfigOverrides.ts:46]: nice! Will be nice to clean this up later on.

File: packages/driver/src/cy/testConfigOverrides.ts
Comment (flotwig) [inline packages/driver/src/cy/testConfigOverrides.ts:128]: If this is what's causing the `driver` tests to hang, it's probably worth rewriting this as recursive instead of a loop. We try to opt for recursion over looping wherever possible because eventually it will hit max stack depth and error, whereas an infinite loop just causes this freezing behavior that's harder to debug. Rewriting it as recursive means if there's another bug at least we get a stack trace.

**Review by flotwig — CHANGES_REQUESTED** [review]
LGTM once test failures are resolved

File: packages/driver/src/cy/testConfigOverrides.ts
Comment (emilyrohrbough) [reply packages/driver/src/cy/testConfigOverrides.ts:127]: This wasn't what was causing the tests to hang. What was causing our tests to hang was the fact the driver tests has a global `beforeEach` hook and a test that set baseUrl to an empty string (ironic given the issue that started this work 😄 ). What was happening was the configuration validation was executing and throwing the error outside of the test context. This caused `driver/src/cypress/cy` to capture the error as an onTop error. When this error is thrown, the runner essential hangs & the test doesn't finish. CircleCi was eventually timing out.

File: packages/driver/src/cy/commands/cookies.ts
Comment (flotwig) [inline packages/driver/src/cy/commands/cookies.ts:163]: was this causing a bug? since before this condition would always pass

**Review by flotwig — APPROVED** [review]

**Comment by flotwig** [comment]
@emilyrohrbough this is breaking, so the base branch should be updated to `9.0.0-release`

**Review by chrisbreiding — APPROVED** [review]

**Review by flotwig — APPROVED** [review]

**Pull request merged**
